File: modules/system/src/main/java/com/jezetek/modules/system/service/PostService.java
package com.jezetek.modules.system.service;

import com.jezetek.core.runtime.BusinessException;
import com.jezetek.modules.system.model.Fetchers;
import com.jezetek.modules.system.model.Post;
import com.jezetek.modules.system.repository.PostRepository;
import com.jezetek.modules.system.repository.UserRepository;
import com.jezetek.modules.system.service.dto.PostInput;
import com.jezetek.modules.system.service.dto.PostSpecification;
import org.babyfish.jimmer.client.FetchBy;
import org.babyfish.jimmer.spring.model.SortUtils;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.security.access.prepost.PreAuthorize;
import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

/*
 * 参照 jimmer-sql 示例的做法：富客户端时代 Controller 层意义弱化，
 * 直接把 web 注解放在 service 上，避免模板代码过度分层
 */
@RestController
@RequestMapping("/api/v1/post")
@Transactional
public class PostService implements Fetchers {

    private final PostRepository postRepository;

    private final UserRepository userRepository;

    public PostService(PostRepository postRepository, UserRepository userRepository) {
        this.postRepository = postRepository;
        this.userRepository = userRepository;
    }

    @PreAuthorize("@perm.has('system:post:list')")
    @GetMapping("/list/bySuperQBE")
    public Page<@NotNull @FetchBy("DEFAULT_FETCHER") Post> findPostsBySuperQBE(
            @RequestParam(defaultValue = "0") int pageIndex,
            @RequestParam(defaultValue = "5") int pageSize,
            @RequestParam(defaultValue = "sortOrder asc") String sortCode,
            PostSpecification specification
    ) {
        return postRepository.find(
                PageRequest.of(pageIndex, pageSize, SortUtils.toSort(sortCode)),
                specification,
                DEFAULT_FETCHER
        );
    }

    @PreAuthorize("@perm.has('system:post:list')")
    @GetMapping("/{id}")
    @Nullable
    public @FetchBy("DEFAULT_FETCHER") Post findPost(
            @PathVariable("id") long id
    ) {
        return postRepository.findById(id, DEFAULT_FETCHER);
    }

    @PreAuthorize("@perm.hasAny('system:post:add', 'system:post:edit')")
    @PutMapping
    public @FetchBy("DEFAULT_FETCHER") Post savePost(
            @Valid @RequestBody PostInput input
    ) {
        return postRepository
                .saveCommand(input)
                .execute(DEFAULT_FETCHER)
                .getModifiedEntity();
    }

    /**
     * 生命周期约束：岗位被用户绑定时禁止删除；
     * 禁用不追溯，已担任该岗位的用户照常生效
     */
    @PreAuthorize("@perm.has('system:post:delete')")
    @DeleteMapping("/{id}")
    public void deletePost(@PathVariable("id") long id) {
        if (userRepository.existsByPostId(id)) {
            throw new BusinessException("岗位已被用户绑定，请先调整用户岗位");
        }
        postRepository.deleteById(id);
    }

    /**
     * 默认抓取形状：全部标量属性
     */
    private static final Fetcher<Post> DEFAULT_FETCHER =
            POST_FETCHER
                    .allScalarFields();
}

File: web/src/api/__generated/services/PostService.ts
import type {Executor} from '../';
import type {PostDto} from '../model/dto/';
import type {Page, PostInput, PostSpecification} from '../model/static/';

export class PostService {
    
    constructor(private executor: Executor) {}
    
    /**
     * 生命周期约束：岗位被用户绑定时禁止删除；
     * 禁用不追溯，已担任该岗位的用户照常生效
     */
    readonly deletePost: (options: PostServiceOptions['deletePost']) => Promise<
        void
    > = async(options) => {
        let _uri = '/api/v1/post/';
        _uri += encodeURIComponent(options.id);
        return (await this.executor({uri: _uri, method: 'DELETE'})) as Promise<void>;
    }
    
    readonly findPost: (options: PostServiceOptions['findPost']) => Promise<
        PostDto['PostService/DEFAULT_FETCHER'] | undefined
    > = async(options) => {
        let _uri = '/api/v1/post/';
        _uri += encodeURIComponent(options.id);
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<PostDto['PostService/DEFAULT_FETCHER'] | undefined>;
    }
    
    readonly findPostsBySuperQBE: (options: PostServiceOptions['findPostsBySuperQBE']) => Promise<
        Page<PostDto['PostService/DEFAULT_FETCHER']>
    > = async(options) => {
        let _uri = '/api/v1/post/list/bySuperQBE';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.specification.keyword;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'keyword='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.pageIndex;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'pageIndex='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.pageSize;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'pageSize='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.sortCode;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'sortCode='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<Page<PostDto['PostService/DEFAULT_FETCHER']>>;
    }
    
    readonly savePost: (options: PostServiceOptions['savePost']) => Promise<
        PostDto['PostService/DEFAULT_FETCHER']
    > = async(options) => {
        let _uri = '/api/v1/post';
        return (await this.executor({uri: _uri, method: 'PUT', body: options.body})) as Promise<PostDto['PostService/DEFAULT_FETCHER']>;
    }
}

export type PostServiceOptions = {
    'findPostsBySuperQBE': {
        readonly pageIndex?: number | undefined, 
        readonly pageSize?: number | undefined, 
        readonly sortCode?: string | undefined, 
        readonly specification: PostSpecification
    }, 
    'findPost': {
        readonly id: number
    }, 
    'savePost': {
        readonly body: PostInput
    }, 
    'deletePost': {
        readonly id: number
    }
}
